compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64RedundantInstructionElimination.java

```diff
@@ -0,0 +1,190 @@
+import org.graalvm.compiler.lir.InstructionValueConsumer;
+import org.graalvm.compiler.lir.LIR;
+import org.graalvm.compiler.lir.LIRInstruction;
+import org.graalvm.compiler.lir.Variable;
+import org.graalvm.compiler.lir.VariableMap;
+import org.graalvm.compiler.lir.gen.LIRGenerationResult;
+import org.graalvm.compiler.lir.phases.PreAllocationOptimizationPhase;
+
+import jdk.vm.ci.code.TargetDescription;
+import jdk.vm.ci.meta.Value;
+
+/**
+ * This optimization tries to remove unused instructions. Currently it only removes the constant
+ * load instructions. The unused constant mov may exist after match rules.
+ */
+public final class AArch64RedundantInstructionElimination extends PreAllocationOptimizationPhase {
+
+    @Override
+    protected void run(TargetDescription target, LIRGenerationResult lirGenRes, PreAllocationOptimizationContext context) {
+        new Optimization(lirGenRes.getLIR()).apply();
+    }
+
+    private static final class Optimization {
+        private final LIR lir;
+        private final VariableMap<DefEntry> unUsedDefs;
+        private final DebugContext debug;
+
+        private Optimization(LIR lir) {
+            this.lir = lir;
+            this.debug = lir.getDebug();
+            this.unUsedDefs = new VariableMap<>();
+        }
+
+        @SuppressWarnings("try")
+        private void apply() {
+            try (Indent indent = debug.logAndIndent("AArch64RedundantInstructionElimination")) {
+                try (DebugContext.Scope s = debug.scope("CollectUnusedInstructions")) {
+                    for (AbstractBlockBase<?> b : lir.getControlFlowGraph().getBlocks()) {
+                        collectUnusedInBlock(b);
+                    }
+
+                    if (!unUsedDefs.isEmpty()) {
+                        debug.log("Redundant instructions that will be removed later: %s", unUsedDefs);
+                    }
+                } catch (Throwable e) {
+                    throw debug.handle(e);
+                }
+
+                try (DebugContext.Scope s = debug.scope("DeleteUnusedInstructions")) {
+                    unUsedDefs.forEach(this::deleteInstruction);
+                    // Remove unused instructions for each block.
+                    for (AbstractBlockBase<?> b : lir.getControlFlowGraph().getBlocks()) {
+                        removeUnusedInBlock(b);
+                    }
+                } catch (Throwable e) {
+                    throw debug.handle(e);
+                }
+            }
+        }
+
+        /**
+         * Collect unused instructions for a {@code block}.
+         */
+        private void collectUnusedInBlock(AbstractBlockBase<?> block) {
+            InstructionValueConsumer defConsumer = (instruction, value, mode, flags) -> {
+                if (isVariable(value) && isEligibleOp(instruction)) {
+                    DefEntry def = new DefEntry(block, instruction, value);
+                    unUsedDefs.put(((Variable) value), def);
```

Comment:
We usually use `LIRValueUtil.asVariable(value)` instead of a direct cast, often with a static import so the code would change to `unUsedDefs.put(asVariable(value)), def);`. But it is a question of style preference and not a blocker.

Reply:
Godd catch! I will update it soon! Thank you!